src/main/java/constants/Regex.java

```diff
@@ -18,7 +21,56 @@ public class Regex {
+    /**
+     * Pattern to match quoted strings or individual words
+     */
     // public static final Pattern USER_COMMAND_REGEX = Pattern.compile("\"[^\"]*\"|\\S+");
     public static final Pattern USER_COMMAND_REGEX = Pattern.compile("\"[^\"]+\"|\\S+");
+
+    //@@author xenthm
+    /**
+     * Regex pattern to extract the author name out of a given input
+     */
+    public static final Pattern AUTHOR_NAME_EXTRACT0R_PATTERN = generateExtractorPattern(AUTHOR_OPTION, SALES_OPTION,
+            BY_DATE_OPTION);
+
+    //@@author xenthm 
+    /**
+     * This method automatically generates a valid regex pattern to use with a field (e.g. author name) extractor. The
+     * regex works by looking for the <code>includedOptionFlag</code> within the input and taking every character after
+     * that as a possibly valid part of the field of interest. If any <code>excludedOptionFlags</code> are found
+     * after the <code>includedOptionFlag</code>, the field of interest is terminated.
+     * <p>
+     * Developers should write their own logic if their <code>Command</code> only has an
+     * <code>includedOptionFlag</code> and no <code>excludedOptionFlags</code> as the parsing logic does not require
+     * this complex regex!
+     *
+     * @param includedOptionFlag  the option flag preceding a field of interest
+     * @param excludedOptionFlags at least one option flag to be excluded
+     * @return compiled regex <code>Pattern</code> to be used with a regex <code>Matcher</code>
+     */
+    private static Pattern generateExtractorPattern(String includedOptionFlag, String... excludedOptionFlags) {
+        assert excludedOptionFlags.length != 0
+                : "Must at least have one excluded option when generating extractor pattern";
+
+        StringBuilder regex = new StringBuilder();
+
+        regex.append("(?<=\\s").append(includedOptionFlag).append(")$");    // The input ends with includedOptionFlag
+        regex.append("|");                                                  // or
+        regex.append("(?<=\\s").append(includedOptionFlag).append("\\s)");  // (there exists includedOptionFlag before
+        regex.append(".*?");                                                // any number of any character
+        regex.append("(?=");                                                // and what comes after them is either:
+        {
+            regex.append("(?<=\\s)");                                       // [any of the excludedOptionFlags
+            regex.append("-[");
+            for (String s : excludedOptionFlags) {
+                regex.append(s.replace("-", EMPTY_REGEX));
+            }
+            regex.append("]");
+            regex.append("(?:\\s|$)");                                      // followed by either a space or the end]
+        }
+        regex.append("|$)");                                                // or the end of the input)
+
+        return Pattern.compile(regex.toString());
+    }
```

Comment:
Really interesting and smart way to use regex. I didn't know we could do this! Will look into adopting this method for our other commands